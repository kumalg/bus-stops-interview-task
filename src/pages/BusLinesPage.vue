<template>
  <div class="bus-lines-page">
    <BaseCard class="bus-line-select-card" title="Select Bus Line">
      <ul class="bus-lines-list">
        <li v-for="line in lines" :key="line" class="bus-line-list-item" @click="selectLine(line)">
          <BaseButton type="button" :is-active="selectedLine === line">{{ line }}</BaseButton>
        </li> 
      </ul> 
    </BaseCard>
    
    <BaseCardEmpty
      v-if="!selectedLine"
      message="Please select the bus line first"
    />
    <BaseCard v-else :title="`Bus Line: ${selectedLine}`" :without-paddings="true">
      <p class="list-title">
        Bus Stops
        <ArrowBottomIcon
          class="order-icon"
          :class="{flip: selectedLineStopsOrder === 'desc' }"
          @click="toggleStopsOrder()"
        />
      </p>
      <ul class="list">
        <li v-for="stop in orderedStops" :key="stop" @click="selectLineStop(stop)">
          {{ stop }}
        </li>
      </ul>
    </BaseCard>

    <BaseCardEmpty
      v-if="!selectedLine || !selectedLineStop"
      :message="!selectedLine ? 'Please select the bus line first' : 'Please select the bus stop first'"
    />
    <BaseCard v-else :title="`Bus Stop: ${selectedLineStop}`" :without-paddings="true">
      <p class="list-title">Time</p>
      <ul class="list">
        <li v-for="time in selectedLineStopTimes" :key="time">{{ time }}</li>
      </ul>
    </BaseCard>
  </div>
</template>

<script setup lang="ts">
import type { Order } from '@/types';

import { computed,  ref } from 'vue';
import { useStore } from 'vuex';

import { toSorted } from '@/assets/sort';

import BaseButton from '@/components/BaseButton.vue';
import BaseCard from '@/components/BaseCard.vue'
import BaseCardEmpty from '@/components/BaseCardEmpty.vue';

import ArrowBottomIcon from '@/components/ArrowBottomIcon.vue'

const store = useStore()
const lines = computed(() => store.getters.lines)

const selectedLine = ref<null | number>(null)
const selectedLineStops = ref<string[]>([])
const selectedLineStopsOrder = ref<Order>('asc')
const selectedLineStop = ref<null | string>(null)
const selectedLineStopTimes = ref<string[]>([])

const orderedStops = computed(() => toSorted(selectedLineStops.value, selectedLineStopsOrder.value))

const selectLine = async (line: number) => {
  selectedLine.value = line
  selectedLineStop.value = null
  selectedLineStops.value = await store.dispatch('getLineStops', line)
}

const selectLineStop = async (stop: string) => {
  selectedLineStop.value = stop
  selectedLineStopTimes.value = await store.dispatch('getTimesForLineStop', { line: selectedLine.value, stop })
}

const toggleStopsOrder = () => {
  selectedLineStopsOrder.value = selectedLineStopsOrder.value === 'asc' ? 'desc' : 'asc'
}
</script>

<style lang="scss" scoped>
.bus-lines-page {
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-template-rows: auto 1fr;
  gap: 1rem;
  height: 100%;

  .bus-line-select-card {
    grid-column: 1 / span 2;
  }

  .bus-lines-list {
    list-style: none;
    padding: 0;
    margin: 0;
    display: flex;
    gap: 0.5rem;
    flex-wrap: wrap;
  }

  .list-title {
    font-size: 0.75rem;
    line-height: 1rem;
    letter-spacing: 0.025em;
    font-weight: 600;
    display: flex;
    gap: 0.25rem;
    color: $color-gray-4;
    margin: 0;
    padding: $card-padding;
    border-bottom: 1px solid $color-gray-15;
  }

  .order-icon {
    color: $color-gray-3;
    cursor: pointer;

    @include default-transition('color, transform');

    &:hover {
      color: $color-gray-4;
    }

    &.flip {
      transform: rotate(180deg);
    }
  }

  .list {
    overflow: auto;
  }
}
</style>