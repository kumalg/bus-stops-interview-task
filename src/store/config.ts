import type { BusStop, BusStopsFetchStatus } from '@/types';

export enum StoreMutation {
    SetBusStops = 'SET_BUS_STOPS',
    SetStatus = 'SET_STATUS',
    SetLineStops = 'SET_LINE_STOPS',
    SetLineStopTimes = 'SET_LINE_STOP_TIMES'
}

export enum StoreAction {
    FetchBusStops = 'fetchBusStops',
    GetLineStops = 'getLineStops',
    GetTimesForLineStop = 'getTimesForLineStop'
}

export type StoreState = {
    busStops: BusStop[];
    status: BusStopsFetchStatus;
    lineStops: {
        [key in number]: BusStop['stop'][];
    };
    lineStopTimes: Map<{ line: BusStop['line']; stop: BusStop['stop'] }, BusStop['time'][]>;
};
