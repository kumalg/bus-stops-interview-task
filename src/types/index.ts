export type BusStop = {
  line: number,
  stop: string,
  order: number,
  time: string
}

export type Order = 'asc' | 'desc'