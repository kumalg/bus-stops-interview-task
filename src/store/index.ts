import type { BusStop } from '@/types';
import type { StoreState } from '@/store/config';

import { createStore, useStore as baseUseStore, Store } from 'vuex';
import { InjectionKey } from 'vue';

import { toSorted, toSortedTimes } from '@/assets/sort';
import { validateBusStops } from '@/assets/validator';

import { StoreAction, StoreMutation } from '@/store/config';

import { BusStopsFetchStatus } from '@/types';

export const key: InjectionKey<Store<StoreState>> = Symbol();

export const store = createStore<StoreState>({
    state: {
        busStops: [],
        status: BusStopsFetchStatus.UNFETCHED,
        lineStops: {},
        lineStopTimes: new Map()
    },
    getters: {
        lines: (state) => toSorted(new Set(state.busStops.map(({ line }) => line))),
        stops: (state) => toSorted(new Set(state.busStops.map(({ stop }) => stop)))
    },
    mutations: {
        [StoreMutation.SetBusStops](state, value: BusStop[]) {
            state.busStops = value;
        },
        [StoreMutation.SetStatus](state, value: BusStopsFetchStatus) {
            state.status = value;
        },
        [StoreMutation.SetLineStops](
            state,
            { line, stops }: { line: BusStop['line']; stops: BusStop['stop'][] }
        ) {
            state.lineStops[line] = stops;
        },
        [StoreMutation.SetLineStopTimes](
            state,
            {
                key,
                times
            }: { key: { line: BusStop['line']; stop: BusStop['stop'] }; times: BusStop['time'][] }
        ) {
            state.lineStopTimes.set(key, times);
        }
    },
    actions: {
        async [StoreAction.FetchBusStops]({ state, commit }) {
            if (state.status === BusStopsFetchStatus.FETCHING) {
                return;
            }

            commit(StoreMutation.SetStatus, BusStopsFetchStatus.FETCHING);

            try {
                const response = await fetch('http://localhost:3000/stops');

                if (!response.ok) {
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
                }

                const json = await response.json();

                if (!validateBusStops(json)) {
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
                } else {
                    commit(StoreMutation.SetBusStops, json);
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.FETCHED);
                }
            } catch {
                commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
            }
        },
        [StoreAction.GetLineStops]({ state, commit }, line: number) {
            if (line in state.lineStops && typeof state.lineStops[line] === 'object') {
                return state.lineStops[line];
            }

            const stopFullObjects = state.busStops.filter((busStop) => busStop.line === line);
            stopFullObjects.sort(({ order: aOrder }, { order: bOrder }) => aOrder - bOrder);

            const stops = [...new Set(stopFullObjects.map(({ stop }) => stop))];

            commit(StoreMutation.SetLineStops, { line, stops });

            return stops;
        },
        [StoreAction.GetTimesForLineStop](
            { state, commit },
            key: { line: BusStop['line']; stop: BusStop['stop'] }
        ) {
            if (state.lineStopTimes.has(key)) {
                return state.lineStopTimes.get(key);
            }

            let times;
            try {
                times = toSortedTimes(
                    new Set(
                        state.busStops
                            .filter(({ line, stop }) => line === key.line && stop === key.stop)
                            .map(({ time }) => time)
                    )
                );
            } catch {
                return [];
            }

            commit(StoreMutation.SetLineStopTimes, { key, times });

            return times;
        }
    },
    modules: {}
});

export const useStore = () => {
    return baseUseStore(key);
};
