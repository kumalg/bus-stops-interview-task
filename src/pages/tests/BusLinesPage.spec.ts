import { flushPromises, shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';

import { BusStopsFetchStatus } from '@/types';

import { StoreAction, StoreState } from '@/store/config';
import { key } from '@/store';

import BusLinesPage from '@/pages/BusLinesPage.vue';

const pushMock = jest.fn();

jest.mock('vue-router', () => ({
    useRouter: jest.fn(() => ({
        push: pushMock,
        currentRoute: {
            value: {
                fullPath: '/',
                query: {}
            }
        }
    }))
}));

const BaseCardStub = {
    template: '<div class="base-card-stub"><slot/></div>'
};

const BaseButtonStub = {
    template: '<button class="base-button-stub"><slot/></button>'
};

const BaseCardEmptyStub = {
    template: '<div class="base-card-empty-stub">{{ message }}</div>',
    props: {
        message: String
    }
};

const CardListStub = {
    template: '<div class="card-list-stub" />'
};

describe('BusLinesPage', () => {
    const LINES = [101, 102, 103] as const;
    const LINE_STOPS: { [key in (typeof LINES)[number]]: string[] } = {
        101: ['stop1', 'stop2', 'stop3'],
        102: ['stop4', 'stop5', 'stop6'],
        103: ['stop7', 'stop8', 'stop9']
    };
    const LINE_STOP_TIMES: StoreState['lineStopTimes'] = new Map();
    LINE_STOP_TIMES.set({ line: 101, stop: LINE_STOPS[101][0] }, ['9:10', '9:15', '9:20']);
    LINE_STOP_TIMES.set({ line: 102, stop: LINE_STOPS[102][0] }, ['10:10', '10:15', '10:20']);
    LINE_STOP_TIMES.set({ line: 103, stop: LINE_STOPS[103][0] }, ['11:10', '11:15', '11:20']);

    const BusLinesPageFactory = () => {
        const state: StoreState = {
            busStops: [],
            status: BusStopsFetchStatus.UNFETCHED,
            lineStops: LINE_STOPS,
            lineStopTimes: LINE_STOP_TIMES
        };

        const store = createStore({
            state,
            getters: {
                lines: () => LINES
            },
            actions: {
                [StoreAction.GetLineStops]() {
                    return LINE_STOPS[101];
                },
                [StoreAction.GetTimesForLineStop]() {
                    return LINE_STOP_TIMES.get({ line: 101, stop: LINE_STOPS[101][0] });
                }
            }
        });
        return shallowMount(BusLinesPage, {
            global: {
                provide: {
                    [key as unknown as string]: store
                },
                stubs: {
                    BaseCard: BaseCardStub,
                    BaseButton: BaseButtonStub,
                    BaseCardEmpty: BaseCardEmptyStub,
                    CardList: CardListStub
                }
            }
        });
    };

    it('renders proper html markup', () => {
        const wrapper = BusLinesPageFactory();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders proper html markup when line and stop selected', async () => {
        const wrapper = BusLinesPageFactory();

        await wrapper.findAll('.line-button')[0]?.trigger('click');

        wrapper
            .findAllComponents(CardListStub)[0]
            .vm.$emit('update:model-value', LINE_STOPS[101][0]);

        await flushPromises();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders correct lines list', () => {
        const wrapper = BusLinesPageFactory();

        const linesElements = wrapper.findAll('.line-button').map((line) => line.text());

        expect(linesElements).toEqual(LINES.map((line) => line.toString()));
    });

    it('renders lines list with empty stops and times if no line selected', () => {
        const wrapper = BusLinesPageFactory();

        expect(wrapper.find('.bus-lines-card').exists()).toBe(true);

        expect(wrapper.find('.bus-stops-empty-card').exists()).toBe(true);
        expect(wrapper.find('.bus-stops-empty-card').text()).toBe(
            'Please select the bus line first'
        );
        expect(wrapper.find('.bus-stops-card').exists()).toBe(false);

        expect(wrapper.find('.bus-times-empty-card').exists()).toBe(true);
        expect(wrapper.find('.bus-times-empty-card').text()).toBe(
            'Please select the bus line first'
        );
        expect(wrapper.find('.bus-times-card').exists()).toBe(false);
    });

    it('renders lines list with stops and empty times if line selected and no stop', async () => {
        const wrapper = BusLinesPageFactory();

        expect(wrapper.find('.bus-lines-card').exists()).toBe(true);

        await wrapper.findAll('.line-button')[0]?.trigger('click');

        expect(wrapper.find('.bus-stops-empty-card').exists()).toBe(false);
        expect(wrapper.find('.bus-stops-card').exists()).toBe(true);

        expect(wrapper.find('.bus-times-empty-card').exists()).toBe(true);
        expect(wrapper.find('.bus-times-empty-card').text()).toBe(
            'Please select the bus stop first'
        );
        expect(wrapper.find('.bus-times-card').exists()).toBe(false);
    });

    it('renders lines list with stops and times if line and stop selected', async () => {
        const wrapper = BusLinesPageFactory();

        expect(wrapper.find('.bus-lines-card').exists()).toBe(true);

        await wrapper.findAll('.line-button')[0]?.trigger('click');

        expect(wrapper.find('.bus-stops-empty-card').exists()).toBe(false);
        expect(wrapper.find('.bus-stops-card').exists()).toBe(true);

        wrapper
            .findAllComponents(CardListStub)[0]
            .vm.$emit('update:model-value', LINE_STOPS[101][0]);

        await flushPromises();

        expect(wrapper.find('.bus-times-empty-card').exists()).toBe(false);
        expect(wrapper.find('.bus-times-card').exists()).toBe(true);
    });
});
