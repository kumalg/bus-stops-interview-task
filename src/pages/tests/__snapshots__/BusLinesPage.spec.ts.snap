// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusLinesPage renders proper html markup 1`] = `
<div
  class="bus-lines-page"
>
  <div
    class="base-card-stub bus-lines-card"
    title="Select Bus Line"
  >
    
    <ul
      class="bus-lines-list"
    >
      
      <li>
        <button
          class="base-button-stub line-button"
          is-active="false"
          type="button"
        >
          
          101
          
        </button>
      </li>
      <li>
        <button
          class="base-button-stub line-button"
          is-active="false"
          type="button"
        >
          
          102
          
        </button>
      </li>
      <li>
        <button
          class="base-button-stub line-button"
          is-active="false"
          type="button"
        >
          
          103
          
        </button>
      </li>
      
    </ul>
    
  </div>
  <transition-stub
    appear="false"
    css="true"
    mode="out-in"
    name="fade"
    persisted="false"
  >
    <div
      class="base-card-empty-stub bus-stops-empty-card"
    >
      Please select the bus line first
    </div>
  </transition-stub>
  <transition-stub
    appear="false"
    css="true"
    mode="out-in"
    name="fade"
    persisted="false"
  >
    <div
      class="base-card-empty-stub bus-times-empty-card"
    >
      Please select the bus line first
    </div>
  </transition-stub>
</div>
`;

exports[`BusLinesPage renders proper html markup when line and stop selected 1`] = `
<div
  class="bus-lines-page"
>
  <div
    class="base-card-stub bus-lines-card"
    title="Select Bus Line"
  >
    
    <ul
      class="bus-lines-list"
    >
      
      <li>
        <button
          class="base-button-stub line-button"
          is-active="true"
          type="button"
        >
          
          101
          
        </button>
      </li>
      <li>
        <button
          class="base-button-stub line-button"
          is-active="false"
          type="button"
        >
          
          102
          
        </button>
      </li>
      <li>
        <button
          class="base-button-stub line-button"
          is-active="false"
          type="button"
        >
          
          103
          
        </button>
      </li>
      
    </ul>
    
  </div>
  <transition-stub
    appear="false"
    css="true"
    mode="out-in"
    name="fade"
    persisted="false"
  >
    <div
      class="card-list-stub bus-stops-card"
      items="stop1,stop2,stop3"
      model-value="stop1"
      reversible="true"
      selectable="true"
      subtitle="Bus Stops"
      title="Bus Line: 101"
    />
  </transition-stub>
  <transition-stub
    appear="false"
    css="true"
    mode="out-in"
    name="fade"
    persisted="false"
  >
    <div
      class="card-list-stub bus-times-card"
      subtitle="Time"
      title="Bus Stop: stop1"
    />
  </transition-stub>
</div>
`;
